### PR TITLE
fix(lsp): display initialization errors

### LINKED ISSUE #25375: Errors in `on_init` and `on_attach` callbacks are not being reported to the user
### Problem

Hello neovim team, thanks so much for all your hard work!

I lost a few hours trying to figure out why my lsp config wasn't being used. Turns out the init callback was throwing an error and neovim did not give any indication. 

I did some digging myself and found the problem. The `on_init` callback is called [here](https://github.com/neovim/neovim/blob/de9348978999aa78f8351b5f55930bb109e742f7/runtime/lua/vim/lsp.lua#L1441). There's a `pcall` and the error is ostensibly passed off to a handler via another `pcall`, but that handler doesn't seem to exist and the return value is dropped.

Unfortunately, `pcall(nil)` doesn't throw and only returns `false, "attempt to call a nil value"`  so both errors are dropped and everyone is none the wiser. 😱 This is a good reminder that discarding the return of `pcall` is a _very bad_ idea. Too risky!

### Steps to reproduce using "nvim -u minimal_init.lua"

```lua
local pattern = 'lua'
local cmd = {'lua-language-server'}
-- Add files/folders here that indicate the root of a project
local root_markers = {'.git', '.editorconfig'}
-- Change to table with settings if required
local settings = vim.empty_dict()

vim.api.nvim_create_autocmd('FileType', {
  pattern = pattern,
  callback = function(args)
    local match = vim.fs.find(root_markers, { path = args.file, upward = true })[1]
    local root_dir = match and vim.fn.fnamemodify(match, ':p:h') or nil
    vim.lsp.start({
      name = 'bugged-ls',
      cmd = cmd,
      root_dir = root_dir,
      settings = settings,
      on_init = function ()
        print("INIT")
        error("whoops")
      end,
      on_attach = function ()
        print("ATTACH")
        error("my bad")
      end
    })
  end
})
```

### Expected behavior

neovim should indicate somehow that an error has occurred during the callback. 

### Neovim version (nvim -v)

NVIM v0.10.0-dev-1203+g0592fd5e-dirty

### Language server name/version

lua-language-server 3.7.1

### Operating system/version

Ubuntu 20

### Log file

_No response_

### PR DESCRIPTION
**Pull request by MariaSolOs**
Fixes #25375

### CODE REVIEW

**Comment by wookayin** [comment]
@MariaSolOs, thanks for working on this! I was also hit by #25375 some times ago so wanted to chime in and share som thoughts.

Although this does a great job of displaying caught errors, it doesn't show any stacktrace. A fix I had internally and locally (but haven't wrapped up enough yet to submit a PR) makes use of `xpcall` and `vim.notify` to print the full stacktrace (which is more aggressive but arguably more useful)

I suggest we could use a similar approach -- I could also help work on this sometime soon.

File: runtime/lua/vim/lsp.lua [hunk 1]
```diff
@@ -1180,10 +1192,7 @@ function lsp.start_client(config)
   ---@see vim.lsp.rpc.client_errors for possible errors. Use
   ---`vim.lsp.rpc.client_errors[code]` to get a human-friendly name.
   function dispatch.on_error(code, err)
-    if log.error() then
-      log.error(log_prefix, 'on_error', { code = lsp.client_errors[code], err = err })
-    end
-    err_message(log_prefix, ': Error ', lsp.client_errors[code], ': ', vim.inspect(err))
+    write_error(code, err)
     if config.on_error then
       local status, usererr = pcall(config.on_error, code, err)
```
Comment (wookayin) [inline runtime/lua/vim/lsp.lua:1197]: How can we document this behavior? The three (or four) additional codes are custom extensions that is not included in `vim.lsp.rpc`.

FYI, the current documentation about `on_error` `|vim.lsp.start_clinet()|` 

```
                  • on_error: Callback with parameters (code, ...), invoked
                    when the client operation throws an error. `code` is a
                    number describing the error. Other arguments may be passed
                    depending on the error kind. See
                    `vim.lsp.rpc.client_errors` for possible errors. Use
                    `vim.lsp.rpc.client_errors[code]` to get human-friendly
                    name.
```

File: runtime/lua/vim/lsp.lua [hunk 2]
```diff
@@ -189,7 +189,9 @@ lsp.client_errors = tbl_extend(
   'error',
   lsp_rpc.client_errors,
   vim.tbl_add_reverse_lookup({
-    ON_INIT_CALLBACK_ERROR = table.maxn(lsp_rpc.client_errors) + 1,
+    BEFORE_INIT_CALLBACK_ERROR = table.maxn(lsp_rpc.client_errors) + 1,
+    ON_INIT_CALLBACK_ERROR = table.maxn(lsp_rpc.client_errors) + 2,
+    ON_ATTACH_ERROR = table.maxn(lsp_rpc.client_errors) + 3,
```
Comment (wookayin) [inline runtime/lua/vim/lsp.lua:194]: missing: ON_EXIT_ERROR (and similar edits to `pcall(config.on_exit, ...)`)

**Comment by MariaSolOs** [comment]
@wookayin thanks for the feedback! For now I'm just following the existing error handling, which involves writing to the log file and error buffer. I'm open to throwing a `vim.notify` as well.

File: runtime/lua/vim/lsp.lua (hunk 1 above)
Comment (MariaSolOs) [reply runtime/lua/vim/lsp.lua:1197]: I'm aware of the documentation of `on_error`, and the way I understand it that function is used when there's an issue once the RPC connection has been established. In this PR I'm dealing with errors that happen before even starting the client, so I'm not sure if that same handler should be used here as well. As for the error codes, there is already an extension to the client codes on master today, but I do agree that they deserve further documentation.

File: runtime/lua/vim/lsp.lua [hunk 3]
```diff
@@ -1172,6 +1174,16 @@ function lsp.start_client(config)
...
   end
 
+  ---Logs the given error to the LSP log and to the error buffer.
+  ---@param code integer Error code
+  ---@param err any Error arguments
+  local function write_error(code, err)
+    if log.error() then
+      log.error(log_prefix, 'on_error', { code = lsp.client_errors[code], err = err })
+    end
+    err_message(log_prefix, ': Error ', lsp.client_errors[code], ': ', vim.inspect(err))
```
[the remark this answers is not in the document]
Comment (MariaSolOs) [reply runtime/lua/vim/lsp.lua:1184]: Given that the error can be anything, I prefer to keep the generic handling that having small nits for specific cases.

File: runtime/lua/vim/lsp.lua [hunk 4]
```diff
@@ -1172,6 +1174,16 @@ function lsp.start_client(config)
     return nil, lsp.rpc_response_error(protocol.ErrorCodes.MethodNotFound)
   end
 
+  ---Logs the given error to the LSP log and to the error buffer.
+  ---@param code integer Error code
+  ---@param err any Error arguments
```
Comment (gpanders) [inline runtime/lua/vim/lsp.lua:1179]: https://github.com/neovim/neovim/pull/25345 was merged so this should now, officially, be `--- `

File: runtime/lua/vim/lsp.lua [hunk 5]
```diff
@@ -1391,8 +1400,10 @@ function lsp.start_client(config)
       trace = valid_traces[config.trace] or 'off',
     }
     if config.before_init then
-      -- TODO(ashkan) handle errors here.
```
Comment (gpanders) [inline runtime/lua/vim/lsp.lua:1394]: Hey, this TODO is finally TODONE! 🥳

File: runtime/lua/vim/lsp.lua (hunk 3 above)
[the remark this answers is not in the document]
Comment (gpanders) [reply runtime/lua/vim/lsp.lua:1184]: It looks like this function is only ever called with the `err` return value from `pcall`? Is that right? AFAIK `pcall` always returns a string for the error message, so if that's the case then `err` here can be assumed to always be a string.

[the remark this answers is not in the document]
Comment (MariaSolOs) [reply runtime/lua/vim/lsp.lua:1184]: @gpanders it's not only used with errors from `pcall`, but also for errors thrown during the RPC communication (see [`dispatch.on_error`](https://github.com/MariaSolOs/neovim/blob/3484beff0564ed6148b35ef5062c434bc2a9aba9/runtime/lua/vim/lsp.lua#L1194)).

File: runtime/lua/vim/lsp.lua (hunk 4 above)
Comment (MariaSolOs) [reply runtime/lua/vim/lsp.lua:1179]: Thanks for pointing this out :)

**Comment by MariaSolOs** [comment]
@mfussenegger I would like to receive your feedback here :)

**Review by justinmk — APPROVED** [review]

**Review by mfussenegger — COMMENTED** [review]
I'm not sure if what we're doing with `client_errors` is sound.

Afaik the rpc client errors are for the documented error codes in the specification (See https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#responseError)

It seems we're conflating those with neovim specific error handling?
But it also looks like this was already the case before, so I guess this is fine - but maybe something to look closer into and change.

(Also for the future, please avoid mixing unrelated style changes into the same commit, it makes reviews more difficult and also increases the chance of conflicts. Ideally you'd make a dedicated PR for that, if we really insist on consistency regarding the new `--- @param` style. I thought the conclusion was to allow both)

**Comment by MariaSolOs** [comment]
> I'm not sure if what we're doing with client_errors is sound.

@mfussenegger You're right. TBH I was just following the existing pattern without stepping back and realizing that this was supposed to be the error object specified by LSP.

> please avoid mixing unrelated style changes into the same commit, it makes reviews more difficult and also increases the chance of conflicts. 

Sorry about that. I will remove the unnecessary whitespace modifications from this PR.

**Comment by MariaSolOs** [comment]
@gpanders @mfussenegger Is there anything else in this PR that you would like me to change?

**Review by gpanders — APPROVED** [review]

**Pull request merged**